### PR TITLE
feat: add basic E2E test

### PR DESCRIPTION
**Pull request by bzp2010**
Add CI-based E2E testing, which mainly includes HTTP server, ETCD server, MySQL data source support, etc.

### CODE REVIEW

**Comment by bzp2010** [comment]
## Update

I have create CI workflow and ginkgo based testing framework. Below is the list of test cases that have been added.

### Test Cases
- [x] MySQL
    - [x] Launch
    - [x] Connect
    - [x] KV Get
- [x] Server
    - [x] HTTP: GET /version

## Suggest and Review

ping @tokers @starsz Please review it and point out the test cases that need to be added if necessary.

File: test/e2e/base/base.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package base
+
+import (
+	"time"
+
+	"github.com/go-resty/resty/v2"
+)
+
+const (
+	DefaultAddress = "127.0.0.1:12379"
```
Comment (tokers) [inline test/e2e/base/base.go:10]: Be strict. Always follow the rules:

1. Do you really need to capitalize the first letter?
2. If so, please add some comments for it.

File: test/e2e/base/run.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	// Determine if the ETCD Adapter started successfully by
+	// checking HTTP accessibility
+	// The default will retry 5 times, HTTP timeout (1s) time with
+	// waiting time (1s) not more than 10 seconds
+	var remainRetries = 5
+	for {
+		if remainRetries == 0 {
+			return nil, errors.New("exceeds the number of run check retries")
+		}
+		resp, err := httpClient.R().Get("http://" + DefaultAddress)
```
Comment (tokers) [inline test/e2e/base/run.go:28]: I suggest using httpexpect here.

File: test/e2e/rdms/rdms_suite_test.go [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+)
+
+var (
+	etcdClient         *clientv3.Client
+	etcdAdapterProcess *exec.Cmd
+)
+
+func TestRDMS(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "RDMS Suite")
```
Comment (starsz) [inline test/e2e/rdms/rdms_suite_test.go:19]: RDMS what's the meaning?

File: test/e2e/base/etcd.go [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+import (
+	"time"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+// NewETCDClient create an ETCD client for testing.
+func NewETCDClient() (*clientv3.Client, error) {
+	cli, err := clientv3.New(clientv3.Config{
+		Endpoints:   []string{"127.0.0.1:12379"},
```
Comment (starsz) [inline test/e2e/base/etcd.go:12]: Can we use `DefaultAddress` here ?

File: test/e2e/server/server_suite_test.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+	"testing"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	"github.com/api7/etcd-adapter/test/e2e/base"
+)
+
+var (
+	serverConfigFile, _ = filepath.Abs("../../testdata/config/server.yaml")
```
Comment (starsz) [inline test/e2e/server/server_suite_test.go:15]: We can use goland `embed`  feature

File: test/e2e/rdms/rdms_suite_test.go (hunk 3 above)
Comment (zaunist) [reply test/e2e/rdms/rdms_suite_test.go:19]: +1

File: test/e2e/server/http_test.go [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+	. "github.com/onsi/gomega"
+	"github.com/tidwall/gjson"
+
+	"github.com/api7/etcd-adapter/test/e2e/base"
+)
+
+var _ = Describe("HTTP", func() {
+	It("Get version", func() {
+		resp, err := base.GetHTTPClient().R().Get("http://" + base.DefaultAddress + "/version")
+		Expect(err).To(BeNil())
```
Comment (zaunist) [inline test/e2e/server/http_test.go:14]:
```suggestion
var _ = Describe("HTTP", func() {
	  It("Get version", func() {
		  resp, err := base.GetHTTPClient().R().Get("http://" + base.DefaultAddress + "/version")
		  Expect(err).To(BeNil())
```

Is it possible to leave spaces in line breaks? Keeping the same indent looks strange...

File: test/e2e/base/base.go (hunk 1 above)
Comment (bzp2010) [reply test/e2e/base/base.go:11]: added

File: test/e2e/base/run.go (hunk 2 above)
Comment (bzp2010) [reply test/e2e/base/run.go:28]: I will try it out, I have no experience with httpexpect.

File: test/e2e/rdms/rdms_suite_test.go (hunk 3 above)
Comment (bzp2010) [reply test/e2e/rdms/rdms_suite_test.go:19]: `RDMS` is the `Relational Database Management System`

Now, I think `datasource` is more appropriate and I will modify it.

File: test/e2e/base/etcd.go (hunk 4 above)
Comment (bzp2010) [reply test/e2e/base/etcd.go]: changed

File: test/e2e/server/server_suite_test.go (hunk 5 above)
Comment (bzp2010) [reply test/e2e/server/server_suite_test.go:15]: The program uses `exec.Command` to start another process and set the `-c path` flag for it, and does not use the contents of the file directly, so there is no need to use go embed.

File: test/e2e/server/http_test.go (hunk 6 above)
Comment (bzp2010) [reply test/e2e/server/http_test.go:14]: @zaunist This is the automatic code formatting performed by GoLand and gofmt. Should we maybe not modify it manually?

File: test/e2e/base/base.go (hunk 1 above)
Comment (tokers) [reply test/e2e/base/base.go:11]: OK.

File: test/e2e/rdms/rdms_suite_test.go (hunk 3 above)
Comment (tokers) [reply test/e2e/rdms/rdms_suite_test.go:19]: I suggest using `RDBMS`.

File: test/e2e/server/http_test.go (hunk 6 above)
Comment (zaunist) [reply test/e2e/server/http_test.go]: Make sense

File: test/e2e/rdms/rdms_suite_test.go (hunk 3 above)
Comment (bzp2010) [reply test/e2e/rdms/rdms_suite_test.go]: > I suggest using RDBMS.

@tokers Considering that we may also add support for NoSQL type databases in the future, I think datasource is a bit more appropriate.

File: test/e2e/base/run.go (hunk 2 above)
Comment (bzp2010) [reply test/e2e/base/run.go]: Migrated to httpexpect

**Review by tokers — APPROVED** [review]

**Comment by bzp2010** [comment]
@starsz @zaunist Please re-review it, thanks.

File: test/e2e/rdms/rdms_suite_test.go (hunk 3 above)
Comment (tokers) [reply test/e2e/rdms/rdms_suite_test.go]: Right, agree with you 👍.

File: internal/adapter/etcd.go [hunk 7]
```diff
@@ -217,8 +217,9 @@ func (a *adapter) handleDeleteEvent(ctx context.Context, ev *Event) {
 }
 
 func (a *adapter) showVersion(w http.ResponseWriter, _ *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
 	w.WriteHeader(http.StatusOK)
-	_, err := w.Write([]byte(`{"etcdserver":"3.5.0-pre","etcdcluster":"3.5.0"}`))
+	_, err := w.Write([]byte(`{"etcdserver":"3.5.0","etcdcluster":"3.5.0"}`))
```
Comment (zaunist) [inline internal/adapter/etcd.go:222]: Is this to use a fixed version that will need to be modified each time the ETCD is updated in the future?

Comment (bzp2010) [reply internal/adapter/etcd.go:222]: No need, we just need to use 3.5.0 all the time.

File: test/e2e/server/server_suite_test.go (hunk 5 above)
Comment (starsz) [reply test/e2e/server/server_suite_test.go:31]: OK

**Review by starsz — APPROVED** [review]

**Comment by bzp2010** [comment]
@zaunist Please take a second look and help merge if there is no problem.

cc @tokers

**Review by zaunist — APPROVED** [review]

**Pull request merged**